scripts/build-bundle

```diff
@@ -135,6 +135,82 @@ tar zxvf ${CHARTS_DIR}/nvidia-driver-runtime-${NVIDIA_DRIVER_RUNTIME_CHART_VERSI
+check_addon_chart_version_matching() {
+  echo "charts packed in Harvester repo"
+  ls -alht ${CHARTS_DIR}
+
+  echo "addon template files"
+  ls -alht ${addons_path}/pkg/templates
+
+  for filename in ${addons_path}/pkg/templates/*.yaml; do
+    local tmpfile=/tmp/$(basename ${filename})
+    grep -v "{{" ${filename} > ${tmpfile}
+    local cnt=$(yq '.resources | length' ${tmpfile})
+
+    local i=0
+    while [[ $i -lt $cnt ]] ; do
+      local chart=$(idx=$i yq '.resources[env(idx)].spec.chart' ${tmpfile})
+      local version=$(idx=$i yq '.resources[env(idx)].spec.version' ${tmpfile})
+      echo addon: "$chart" version: $version
+
+      local EXIT_CODE=0
+      local repover=$(chart=$chart yq '.entries[strenv(chart)][0].version' < ${CHARTS_DIR}/index.yaml) || EXIT_CODE=$?
+      if [ $EXIT_CODE != 0 ]; then
+       echo WARNING: addon $chart is defined, but the chart is not packed into repo / repo struct is changed
+       continue
+      fi
+
+      # some charts are not packed into arm64 ISO, the above yq will return `null`
+      if [[ $repover == "null" ]] && [[ ${ARCH} == "arm64" ]]; then
+       echo  WARNING: addon "$chart" is defined with version "$version" but the chart is not packed into repo in ${ARCH}
+      elif [[ $repover != $version ]]; then
+       echo  addon "$chart" has version mis-matching: in repo is "$repover" but in addon is "$version"
+       return 1
+      fi
+      (( i += 1 ))
+    done
+  done
+  echo "all addons have matched version with the repo"
+}
+
+check_addon_chart_version_matching
+
+check_upgrade_addon_chart_version_matching() {
+  echo "upgrade addon files"
+  ls -alht ${harvester_path}/package/upgrade/addons
+
+  for filename in ${harvester_path}/package/upgrade/addons/*.yaml; do
+    local tmpfile=/tmp/$(basename ${filename})
+    grep -v "{{" ${filename} > ${tmpfile}
+
+    local chart=$(yq '.spec.chart' ${tmpfile})
+    local version=$(yq '.spec.version' ${tmpfile})
+    echo upgrade addon: "$chart" version: $version
+
+    local EXIT_CODE=0
+    local repover=$(chart=$chart yq '.entries[strenv(chart)][0].version' < ${CHARTS_DIR}/index.yaml) || EXIT_CODE=$?
+    if [ $EXIT_CODE != 0 ]; then
+     echo WARNING: upgrade addon $chart is defined, but the chart is not packed into repo / repo struct is changed
+     continue
+    fi
+
+    # some charts are not packed into arm64 ISO, the above yq will return `null`
+    if [[ $repover == "null" ]] && [[ ${ARCH} == "arm64" ]]; then
+     echo  WARNING: upgrade addon "$chart" is defined with version "$version" but the chart is not packed into repo in ${ARCH}
+    elif [[ $repover != $version ]]; then
+     # the upgrade addon is stored in GH harvester; but the chart repo is stored in GH harvester-installer
+     # a new version bumping generally means two PRs
+     # directly return 1 will cause failure in build
+     echo  WARNING upgrade addon "$chart" has version mis-matching: in repo is "$repover" but in addon is "$version"
```

Comment:
I think this should return 1 to fail the build if the upgrade addons version doesn't match. I mean, I assume we don't ever want to build an ISO where the repo addon versions don't match the upgrade addon versions?

Reply:
The current consideration is:

When bumping a new version, we need:
(1) PR in harvester-installer to add new chart version to repo
(2) Update the addon manifest definition, originally in harvester-installer, but now in addon repository
(3) Update the addon upgrade in Harvester

Three PRs are needed. Before all are merged, the `build iso` will fail in the middle if we return 1 here, that is a bit not convenient for PR merging.

We may add an option to forcely fail / alternatively continue, and it is used when publishing an ISO.

Reply:
> We may add an option to forcely fail / alternatively continue, and it is used when publishing an ISO.

Perhaps we can check if the harvester version is a tagged version (e.g. v1.3.0), if so, return 1 to fail fast. (Not sure if this is an option, just leave a comment here 😃 )

Reply:
After https://github.com/harvester/harvester/issues/5859 is done, we can return 1 here.